test-helpers/src/suites/binary_proofs.rs

```diff
@@ -0,0 +1,182 @@
+use fuel_merkle::binary::MerkleTree;
+use fuel_merkle::common::{Bytes32, StorageMap};
+
+use fuel_merkle_test_helpers::data::{binary::ProofTest, EncodedValue, Encoding};
+
+use digest::Digest;
+use rand::seq::IteratorRandom;
+use rand::SeedableRng;
+use rand_chacha::ChaCha8Rng;
+use sha2::Sha256;
+use std::convert::TryInto;
+
+type Hash = Sha256;
+
+pub fn sum(data: &[u8]) -> Bytes32 {
+    let mut hash = Hash::new();
+    hash.update(&data);
+    hash.finalize().try_into().unwrap()
+}
+
+fn generate_test(
+    name: String,
+    description: String,
+    sample_data: &Vec<Bytes32>,
+    proof_index: u64,
+) -> ProofTest {
+    let (root, proof_set) = {
+        let storage = StorageMap::new();
+        let mut test_tree = MerkleTree::new(storage);
+        for datum in sample_data.iter() {
+            test_tree.push(datum).unwrap();
+        }
+        // SAFETY: prove(i) is guaranteed to return a valid proof if the proof
+        // index is within the range of valid leaves. proof_index will always
+        // be selected from this range.
+        test_tree.prove(proof_index).unwrap()
+    };
+
+    // SAFETY: All EncodedValues are specified with a valid encoding.
+    let encoded_root = EncodedValue::from_raw(&root, Encoding::Hex);
+    let encoded_proof_set = proof_set
+        .iter()
+        .map(|v| EncodedValue::from_raw(&v, Encoding::Hex))
+        .collect::<Vec<_>>();
+    let num_leaves = sample_data.len() as u64;
+
+    ProofTest {
+        name,
+        description,
+        root: encoded_root,
+        proof_set: encoded_proof_set,
+        proof_index,
+        num_leaves,
+        expected_verification: true,
+    }
+}
+
+fn write_test(test: &ProofTest) {
+    let yaml = serde_yaml::to_string(test).expect("Unable to serialize test!");
+    std::fs::write(
+        format!("../tests-data-binary/fixtures/{}.yaml", test.name),
+        yaml,
+    )
+    .expect("Unable to write file!");
+}
+
+fn main() {
+    let test_data_count = 2u64.pow(16);
+    let test_data = (0..test_data_count)
+        .map(|i| sum(&i.to_be_bytes()))
+        .collect::<Vec<Bytes32>>();
+
+    let mut rng = ChaCha8Rng::seed_from_u64(90210);
```

Comment:
I agree with @freesig , it is hard to read those yaml files hard=) But it is not hard to read a code. The description definitely makes it more clear. 

What do you think about splitting each test case into its inner function and using the function's name also in the description(like "generated by `invalid_index_512`")? To make a ref to the code(anyone can search in the project for the method with the same name).

Also, it can be useful to use its own `rng` in each test case. In this case, if you will add/remove test case, it will not affect other yaml files.

Reply:
> What do you think about splitting each test case into its inner function and using the function's name also in the description(like "generated by invalid_index_512")? To make a ref to the code(anyone can search in the project for the method with the same name).

If I understand correctly, you're suggesting a separate function for each test, something like:
``` rust
fn generate_test_invalid_index_512() { ... }
fn generate_test_invalid_index_1024() { ... }
...

fn main() {
    generate_test_invalid_index_512();
    generate_test_invalid_index_1024();
    ...
}
```

This way, we can reference the test name in the test YAML and readers of the YAML file can review the code that generated the YAML output. 

I think that's fine. I will refactor that. 

However, I don't think this adds much value to consumers of the test suite outside of `fuel-merkle`, like `fuel-merkle-sol`, where the test is implemented in Typescript. I think the value is limited because the consumer would have to:
- Download `fuel-merkle` (or view it online) to see the code that this is referencing
- Be able to read Rust code
- Understand the `fuel-merkle` library, at least somewhat

Ultimately, the value of this test suite (and test suites like it) is to ensure compliance with `fuel-spec`. It isn't a substitute for unit tests or other debugging tools where the developer may need to interact with the tests as they develop the code.

Reply:
> Also, it can be useful to use its own rng in each test case. In this case, if you will add/remove test case, it will not affect other yaml files.

Yes, this is certainly an improvement! I will add that too.